File: solutions/cpp/Palindrome_linked_list.cpp
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>

#include "./Linked_list_prototype.h"
#include "./Reverse_linked_list_iterative.h"

using std::cout;
using std::endl;
using std::make_shared;
using std::shared_ptr;

// @include
bool IsLinkedListAPalindrome(shared_ptr<ListNode<int>> L) {
  shared_ptr<ListNode<int>> slow = L, fast = L;
  while (fast && fast->next) {
    fast = fast->next->next, slow = slow->next;
  }

  // Compares the first half and the reversed second half lists.
  shared_ptr<ListNode<int>> reverse = ReverseLinkedList(slow->next);
  while (reverse && L) {
    if (reverse->data != L->data) {
      return false;
    }
    reverse = reverse->next, L = L->next;
  }
  return true;
}
// @exclude

template <typename T>
void PrintList(shared_ptr<ListNode<T>> L) {
  while (L) {
    cout << L->data << ' ';
    L = L->next;
  }
  cout << endl;
}

int main(int argc, char* argv[]) {
  shared_ptr<ListNode<int>> head = nullptr;
  if (argc > 2) {
    // Input the node's value in reverse order.
    for (int i = 1; i < argc; ++i) {
      shared_ptr<ListNode<int>> curr =
          make_shared<ListNode<int>>(ListNode<int>{atoi(argv[i]), head});
      head = curr;
    }
    cout << ((IsLinkedListAPalindrome(head)) ? "Yes" : "No") << endl;
  } else {
    // A link list is a palindrome.
    for (int i = 6; i >= 1; --i) {
      shared_ptr<ListNode<int>> curr =
          make_shared<ListNode<int>>(ListNode<int>{1, head});
      head = curr;
    }
    assert(IsLinkedListAPalindrome(head) == true);
    // Still a palindrome linked list.
    head = shared_ptr<ListNode<int>>(nullptr);
    for (int i = 5; i >= 1; --i) {
      shared_ptr<ListNode<int>> curr =
          make_shared<ListNode<int>>(ListNode<int>{1, head});
      head = curr;
    }
    head->next->next->data = 3;
    assert(IsLinkedListAPalindrome(head) == true);
    // Not a palindrome linked list.
    head = nullptr;
    for (int i = 5; i >= 1; --i) {
      shared_ptr<ListNode<int>> curr =
          make_shared<ListNode<int>>(ListNode<int>{i, head});
      head = curr;
    }
    assert(IsLinkedListAPalindrome(head) == false);
  }
  return 0;
}

File: solutions/python/palindrome_linked_list.py
import sys
from linked_list_prototype import ListNode
from reverse_linked_list_iterative import reverse_linked_list


# @include
def is_linked_list_a_palindrome(L):
    # Finds the second half of L.
    slow = fast = L
    while fast and fast.next:
        fast, slow = fast.next.next, slow.next

    # Compares the first half and the reversed second half lists.
    first_half_iter, second_half_iter = L, reverse_linked_list(slow)
    while second_half_iter and first_half_iter:
        if second_half_iter.data != first_half_iter.data:
            return False
        second_half_iter, first_half_iter = (second_half_iter.next,
                                             first_half_iter.next)
    return True
# @exclude


def main():
    head = None
    if len(sys.argv) > 2:
        # Input the node's value in reverse order.
        for i in sys.argv[1:]:
            curr = ListNode(int(i), head)
            head = curr
        print('Yes' if is_linked_list_a_palindrome(head) else 'No')
    assert is_linked_list_a_palindrome(None)
    assert is_linked_list_a_palindrome(ListNode(1))
    assert is_linked_list_a_palindrome(ListNode(1, ListNode(1)))
    assert is_linked_list_a_palindrome(ListNode(1, ListNode(2))) == False
    assert is_linked_list_a_palindrome(
        ListNode(1, ListNode(3, ListNode(2, ListNode(1))))) == False

    head = None
    # A link list is a palindrome.
    for _ in range(6):
        curr = ListNode(1, head)
        head = curr
    assert is_linked_list_a_palindrome(head)

    # Still a palindrome linked list.
    head = None
    for _ in range(5):
        curr = ListNode(1, head)
        head = curr
    head.next.next.data = 3
    assert is_linked_list_a_palindrome(head)

    # Not a palindrome linked list.
    head = None
    for i in reversed(range(1, 5)):
        curr = ListNode(i, head)
        head = curr
    assert is_linked_list_a_palindrome(head) == False


if __name__ == '__main__':
    main()
